dao/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,237 @@
+
+    @Autowired
+    public GiftCertificateDaoImpl(JdbcTemplate template, RowMapper<GiftCertificate> rowMapper) {
+        this.template = template;
+        this.rowMapper = rowMapper;
+    }
+
+
+    /**
+     * This method saves GiftCertificate entity.
+     * The method uses KayHolder for getting generated id for GiftCertificate entity from db.
+     *
+     * @param certificate GiftCertificate entity without id.
+     * @return GiftCertificate entity with generated id.
+     * @since 1.0
+     */
+    @Override
+    public GiftCertificate save(GiftCertificate certificate) {
+
+        String name = certificate.getName();
+        String description = certificate.getDescription();
+        BigDecimal price = certificate.getPrice();
+        int duration = certificate.getDuration();
+        String createDate = certificate.getCreateDate();
+        String lastUpdateDate = certificate.getLastUpdateDate();
+
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        template.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(SAVE_QUERY, Statement.RETURN_GENERATED_KEYS);
+            ps.setString(1, name);
+            ps.setString(2, description);
+            ps.setBigDecimal(3, price);
+            ps.setInt(4, duration);
+            ps.setString(5, createDate);
+            ps.setString(6, lastUpdateDate);
+            return ps;
+        }, keyHolder);
+
+        certificate.setId(keyHolder.getKey().longValue());
+
+        return certificate;
+    }
+
+    /**
+     * This method get GiftCertificate entity by id.
+     *
+     * @param id GiftCertificate entity's id.
+     * @return Optional of GiftCertificate entity. If there is no GiftCertificate with given id, return Optional.empty().
+     * @since 1.0
+     */
+    @Override
+    public Optional<GiftCertificate> getById(long id) {
+        GiftCertificate result = template.queryForStream(GET_BY_ID_QUERY, rowMapper, id).findFirst().orElse(null);
+        return Optional.ofNullable(result);
+    }
+
+    /**
+     * This method get all GiftCertificate entities.
+     *
+     * @return List of all GiftCertificate entities.
+     * @since 1.0
+     */
+    @Override
+    public List<GiftCertificate> findAll() {
+        return template.query(GET_ALL_QUERY, rowMapper);
+    }
+
+
+    /**
+     * This method get GiftCertificate entities by part of their description.
+     *
+     * @param partOfDescription part of GiftCertificate entity's description for search
+     * @return List of GiftCertificate entities which was found by part of description.
+     * @since 1.0
+     */
+    @Override
+    public List<GiftCertificate> getAllByPartOfDescription(String partOfDescription) {
+
+        SqlParameterSource params = new MapSqlParameterSource().addValue(NAME_IN_PARAM_FOR_PART_OF_DESCRIPTION, partOfDescription);
+        Map<String, Object> procedureResult = executeProcedure(FIND_BY_DESCRIPTION_PART_PROCEDURE_NAME, params);
```

Comment:
I really do not understand why you just cannot use like '%part%' operator and get rid of your procedures

Reply:
Just following the terms of our module C:
![image](https://user-images.githubusercontent.com/58271266/113395712-f1769e80-93a2-11eb-89a6-6a32e2104826.png)